### PR TITLE
docs: add contributing.md

### PR DESCRIPTION
**Pull request by louis-bompart**
https://coveord.atlassian.net/browse/CDX-1144

## Proposed changes

A long overdue document explains the what and hows of the project.
Whys have been omitted to not rewrite LotR. There's enough Tolkien with AMZ lately.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+## The CLI folder
+
+The CLI folder contains 3 special projects:
+
+- `cli-commons-dev`: a set of development utilities (e.g. eslint config) that is commons between the CLI and its plugins.
+- `cli-commons`: a set of **production** utilities (e.g. error wrapping) that is used across the CLI and its plugins.
+- `core`: The CLI Itself.
+
+This allow the CLI to ship as a bundle of bricks, instead of a big chunky house.
+However, our tools doesn't play necessarly nicely with this structure.
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply CONTRIBUTING.md:22]: I will clarify, but essentially: Oclif. It required quite a few tweaks

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# Contributing to the Coveo CLI
+
+The Coveo CLI repository is organized as a monorepo. To learn more about repository, check-out [monorepo.tools](https://monorepo.tools/).
+Its packages are split into 3 big categories:
```
Comment (jpmarceau) [inline CONTRIBUTING.md:4]:
```suggestion
Its packages are split into three categories:
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+- UI: Under `packages/ui` you'll find the project 'graviting' around the CLI and that helps developer to get started with a Headless implementation.
+- CLI: Under `packages/cli` you'll find the 'vanilla' plugin of the CLI, the CLI itself, and some utils to help with the development of the CLI and its plugins.
+<!-- TODO: Remove when the E2E will finally be 'split' -->
+- E2E: A special 'package' under `packages/cli-e2e`, it contains the end-to-end tests for the CLI and its included plugin.
```
Comment (jpmarceau) [inline CONTRIBUTING.md:9]:
```suggestion
- E2E: A special 'package' under `packages/cli-e2e`, which contains the end-to-end tests for the CLI and its included plugins.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+The rest of this document will focus on the CLI and its plugins.
+
+## The CLI folder
+
+The CLI folder contains 3 special projects:
```
Comment (jpmarceau) [inline CONTRIBUTING.md:15]:
```suggestion
The CLI folder contains three special projects:
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+### The TypeScript Configs
+
+You'll find up to 3 typescript configs in the CLI projects:
+
+- `tsconfig.json`:
+  - Here to serve as a 'local' base configuration for the other, but more importantly
+  - Used as default by the language server of your IDE (think wiggly red lines)
+- `tsconfig.dev.json`:
+  - Only for CLI and plugins, it is used when you run the commands in **dev** mode with OClif
```
Comment (jpmarceau) [inline CONTRIBUTING.md:32]:
```suggestion
  - Only for CLI and plugins. Used when you run the commands in **dev** mode with OClif.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  - It uses the aforementioned references to identify the 'external' bits.
+
+### The package.json files
+
+When checking `cli-commons` and `cli-commons-dev`, you'll see an `exports` fields.
+This give instruction on where to find some files to the package resolver, so that the import interface stay simples and consistent between development and production.
+
+For more info on that, check [NodeJS documentation](https://nodejs.org/api/packages.html#subpath-exports)
+
+All those configurations, one way or another, does point to the `tsconfig.json` of `cli-commons-dev`.
```
[the remark this answers is not in the document]
Comment (jpmarceau) [reply CONTRIBUTING.md:47]: oups I just realized that I may have edited some of Yassine's suggestions by mistake 🙊 
Hopefully I won't have caused confusion 🙃

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
+# Contributing to the Coveo CLI
+
+The Coveo CLI repository is organized as a monorepo. To learn more about repository, check-out [monorepo.tools](https://monorepo.tools/).
+Its packages are split into 3 big categories:
+
+- UI: Under `packages/ui` you'll find the project 'graviting' around the CLI and that helps developer to get started with a Headless implementation.
```
[the remark this answers is not in the document]
Comment (louis-bompart) [reply CONTRIBUTING.md:6]: It's not in master yet it'll be done in #939 
This is why this PR is actually a follow-up on #939

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+## The CLI Folder
+
+The CLI folder contains three special projects:
+
+- `cli-commons-dev`: a set of development utilities (e.g. eslint config) shared across the CLI and its plugins.
+- `cli-commons`: a set of **production** utilities (e.g. error wrapping) shared across the CLI and its plugins.
+- `core`: The CLI Itself.
+
+This structure allows the CLI to ship as a bundle of bricks, rather than as a big chunky house.
+However, our tools don't necessarily play nicely with this structure.
```
Comment (louis-bompart) [inline CONTRIBUTING.md:22]:
```suggestion
However, `oclif` doesn't necessarily play nicely with this project structure
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
+# Contributing to the Coveo CLI
+
+The Coveo CLI repository is organized as a monorepo (see [monorepo.tools](https://monorepo.tools/)).
+Its packages are split into three categories:
+
+- UI: Under `packages/ui` you'll find the project 'gravitating' around the CLI that helps developers to get started with a Headless implementation.
```
Comment (jpmarceau) [inline CONTRIBUTING.md:6]:
```suggestion
- UI: Under `packages/ui` you'll find the project 'gravitating' around the CLI that helps developers get started with a Headless implementation.
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  - To work, it needs:
+    - The `@oclif/core` patch
+    - The custom `dev` node executable
+- `tsconfig.build.json`:
+  - Only used for the transpilation of the shipped packages.
+  - It uses the aforementioned references to identify the 'external' bits.
+
+### The package.json Files
+
+When checking `cli-commons` and `cli-commons-dev`, you'll see an `exports` field.
```
Comment (jpmarceau) [inline CONTRIBUTING.md:42]:
```suggestion
When looking at `cli-commons` and `cli-commons-dev`, you'll see an `exports` field.
```

**Review by jpmarceau — APPROVED** [review]
👍

**Review by ThibodeauJF — APPROVED** [review]

**Comment by ThibodeauJF** [comment]
Compared to AMZ's LoTR, this is a welcomed addition

**Pull request merged**